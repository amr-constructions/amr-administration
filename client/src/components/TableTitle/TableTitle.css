.table-title-heading {
  float: left;
  margin-top: 0.3rem;
}

.table-title-button {
  float: right;
}