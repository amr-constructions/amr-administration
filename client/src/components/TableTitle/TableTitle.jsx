import React from 'react';
import { Button } from 'antd';
import PropTypes from 'prop-types';
import './TableTitle.css';

const TableTitle = ({ title, button }) => {
  let buttonComponent = '';
  const needToRenderButton = (button && (button.label || button.icon));
  if (needToRenderButton) {
    buttonComponent = React.createElement(Button, {
      ...button,
      icon: button.icon ? React.createElement(button.icon) : '',
    }, button.label);
  }

  return (
    <div>
      <h4 className="table-title-heading">{title}</h4>

      {
        (needToRenderButton) ? <h4 className="table-title-button">{ buttonComponent }</h4> : ''
      }

      <div style={{
        clear: 'both',
      }}
      />
    </div>
  );
};

TableTitle.propTypes = {
  title: PropTypes.string.isRequired,
  button: PropTypes.shape({
    icon: PropTypes.elementType,
    type: PropTypes.string,
    label: PropTypes.string,
    onClick: PropTypes.func,
  }),
};

TableTitle.defaultProps = {
  button: null,
};

export default TableTitle;
