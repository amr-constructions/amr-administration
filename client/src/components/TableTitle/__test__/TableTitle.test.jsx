import { render, cleanup, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import React from 'react';
import ReactDOM from 'react-dom';
import { OrderedListOutlined } from '@ant-design/icons';
import renderer from 'react-test-renderer';
import TableTitle from '../TableTitle';

afterEach(() => {
  cleanup();
  jest.clearAllMocks();
});

test('renders without crashing', () => {
  const div = document.createElement('div');

  const component = <TableTitle title="Title Of Table" />;

  ReactDOM.render(component, div);
  ReactDOM.unmountComponentAtNode(div);

  expect(global.console.error).toHaveBeenCalledTimes(0);

  const tree = renderer.create(component).toJSON();
  expect(tree).toMatchSnapshot();
});

test('render fails if no title prop given', () => {
  const div = document.createElement('div');
  const component = <TableTitle />;

  ReactDOM.render(component, div);

  expect(global.console.error).toHaveBeenCalledTimes(1);

  const tree = renderer.create(component).toJSON();
  expect(tree).toMatchSnapshot();
});

test('renders title prop and not the button', () => {
  const component = <TableTitle title="Title Of Table" />;

  const { getByText, queryAllByRole } = render(component);

  const titleElement = getByText(/Title Of Table/, {
    selector: 'h4.table-title-heading',
  });
  expect(titleElement).toBeInTheDocument();

  expect(queryAllByRole('button')).toHaveLength(0);

  expect(global.console.error).toHaveBeenCalledTimes(0);

  const tree = renderer.create(component).toJSON();
  expect(tree).toMatchSnapshot();
});

test('renders title prop and basic button', () => {
  const buttonProps = {
    label: 'My Button',
  };

  const component = (
    <TableTitle
      title="Title Of Table"
      button={buttonProps}
    />
  );

  const { getByText, queryAllByRole } = render(component);

  const titleElement = getByText(/Title Of Table/, {
    selector: 'h4.table-title-heading',
  });
  expect(titleElement).toBeInTheDocument();

  const buttonElement = getByText(buttonProps.label, {
    selector: 'h4.table-title-button button span',
  });
  expect(buttonElement).toBeInTheDocument();

  expect(queryAllByRole('button')).toHaveLength(1);

  expect(global.console.error).toHaveBeenCalledTimes(0);

  const tree = renderer.create(component).toJSON();
  expect(tree).toMatchSnapshot();
});

test('renders title prop and button with onClick handler', () => {
  const buttonProps = {
    label: 'Button With Callbac',
    onClick: jest.fn(),
  };

  const component = (
    <TableTitle
      title="Title Of Table"
      button={buttonProps}
    />
  );

  const { getByText, queryAllByRole } = render(component);

  const titleElement = getByText(/Title Of Table/, {
    selector: 'h4.table-title-heading',
  });
  expect(titleElement).toBeInTheDocument();

  const buttonElement = getByText(buttonProps.label, {
    selector: 'h4.table-title-button button span',
  });
  expect(buttonElement).toBeInTheDocument();

  expect(queryAllByRole('button')).toHaveLength(1);

  fireEvent.click(buttonElement);

  expect(buttonProps.onClick).toHaveBeenCalledTimes(1);

  expect(global.console.error).toHaveBeenCalledTimes(0);

  const tree = renderer.create(component).toJSON();
  expect(tree).toMatchSnapshot();
});

test('invalid icon to button', () => {
  const buttonProps = {
    label: 'Button With Callbac',
    onClick: jest.fn(),
    icon: 'asd',
  };

  const component = (
    <TableTitle
      title="Title Of Table"
      button={buttonProps}
    />
  );

  const { getByText, queryAllByRole } = render(component);

  const titleElement = getByText(/Title Of Table/, {
    selector: 'h4.table-title-heading',
  });
  expect(titleElement).toBeInTheDocument();

  const buttonElement = getByText(buttonProps.label, {
    selector: 'h4.table-title-button button span',
  });
  expect(buttonElement).toBeInTheDocument();

  expect(queryAllByRole('button')).toHaveLength(1);

  fireEvent.click(buttonElement);

  expect(buttonProps.onClick).toHaveBeenCalledTimes(1);

  expect(global.console.error).toHaveBeenCalledTimes(1);

  const tree = renderer.create(component).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Button with all props', () => {
  const buttonProps = {
    label: 'Button With Callbac',
    onClick: jest.fn(),
    icon: OrderedListOutlined,
  };

  const component = (
    <TableTitle
      title="Title Of Table"
      button={buttonProps}
    />
  );

  const { getByText, queryAllByRole } = render(component);

  const titleElement = getByText(/Title Of Table/, {
    selector: 'h4.table-title-heading',
  });
  expect(titleElement).toBeInTheDocument();

  const buttonElement = getByText(buttonProps.label, {
    selector: 'h4.table-title-button button span',
  });
  expect(buttonElement).toBeInTheDocument();

  expect(queryAllByRole('button')).toHaveLength(1);

  fireEvent.click(buttonElement);

  expect(buttonProps.onClick).toHaveBeenCalledTimes(1);

  expect(queryAllByRole('img')).toHaveLength(1);

  expect(global.console.error).toHaveBeenCalledTimes(0);

  const tree = renderer.create(component).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Icon only button', () => {
  const buttonProps = {
    onClick: jest.fn(),
    icon: OrderedListOutlined,
  };

  const { getByText, queryAllByRole } = render(<TableTitle
    title="Title Of Table"
    button={buttonProps}
  />);

  const titleElement = getByText(/Title Of Table/, {
    selector: 'h4.table-title-heading',
  });
  expect(titleElement).toBeInTheDocument();

  const buttonElement = getByText('', {
    selector: 'h4.table-title-button button span',
  });
  expect(buttonElement).toBeInTheDocument();

  expect(queryAllByRole('button')).toHaveLength(1);

  fireEvent.click(buttonElement);
  fireEvent.click(buttonElement);

  expect(buttonProps.onClick).toHaveBeenCalledTimes(2);

  expect(queryAllByRole('img')).toHaveLength(1);

  expect(global.console.error).toHaveBeenCalledTimes(0);

  const tree = renderer.create(<TableTitle
    title="Title Of Table"
    button={buttonProps}
  />).toJSON();

  expect(tree).toMatchSnapshot();
});

global.console = {
  ...global.console,
  error: jest.fn(),
};

Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});
