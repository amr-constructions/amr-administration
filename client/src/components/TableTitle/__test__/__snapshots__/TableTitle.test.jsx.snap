// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button with all props 1`] = `
<div>
  <h4
    className="table-title-heading"
  >
    Title Of Table
  </h4>
  <h4
    className="table-title-button"
  >
    <button
      className="ant-btn"
      label="Button With Callbac"
      onClick={[Function]}
      type="button"
    >
      <span
        aria-label="ordered-list"
        className="anticon anticon-ordered-list"
        role="img"
      >
        <svg
          aria-hidden="true"
          className=""
          data-icon="ordered-list"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M920 760H336c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h584c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0-568H336c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h584c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 284H336c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h584c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM216 712H100c-2.2 0-4 1.8-4 4v34c0 2.2 1.8 4 4 4h72.4v20.5h-35.7c-2.2 0-4 1.8-4 4v34c0 2.2 1.8 4 4 4h35.7V838H100c-2.2 0-4 1.8-4 4v34c0 2.2 1.8 4 4 4h116c2.2 0 4-1.8 4-4V716c0-2.2-1.8-4-4-4zM100 188h38v120c0 2.2 1.8 4 4 4h40c2.2 0 4-1.8 4-4V152c0-4.4-3.6-8-8-8h-78c-2.2 0-4 1.8-4 4v36c0 2.2 1.8 4 4 4zm116 240H100c-2.2 0-4 1.8-4 4v36c0 2.2 1.8 4 4 4h68.4l-70.3 77.7a8.3 8.3 0 00-2.1 5.4V592c0 2.2 1.8 4 4 4h116c2.2 0 4-1.8 4-4v-36c0-2.2-1.8-4-4-4h-68.4l70.3-77.7a8.3 8.3 0 002.1-5.4V432c0-2.2-1.8-4-4-4z"
          />
        </svg>
      </span>
      <span>
        Button With Callbac
      </span>
    </button>
  </h4>
  <div
    style={
      Object {
        "clear": "both",
      }
    }
  />
</div>
`;

exports[`Icon only button 1`] = `
<div>
  <h4
    className="table-title-heading"
  >
    Title Of Table
  </h4>
  <h4
    className="table-title-button"
  >
    <button
      className="ant-btn ant-btn-icon-only"
      onClick={[Function]}
      type="button"
    >
      <span
        aria-label="ordered-list"
        className="anticon anticon-ordered-list"
        role="img"
      >
        <svg
          aria-hidden="true"
          className=""
          data-icon="ordered-list"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M920 760H336c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h584c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0-568H336c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h584c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 284H336c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h584c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM216 712H100c-2.2 0-4 1.8-4 4v34c0 2.2 1.8 4 4 4h72.4v20.5h-35.7c-2.2 0-4 1.8-4 4v34c0 2.2 1.8 4 4 4h35.7V838H100c-2.2 0-4 1.8-4 4v34c0 2.2 1.8 4 4 4h116c2.2 0 4-1.8 4-4V716c0-2.2-1.8-4-4-4zM100 188h38v120c0 2.2 1.8 4 4 4h40c2.2 0 4-1.8 4-4V152c0-4.4-3.6-8-8-8h-78c-2.2 0-4 1.8-4 4v36c0 2.2 1.8 4 4 4zm116 240H100c-2.2 0-4 1.8-4 4v36c0 2.2 1.8 4 4 4h68.4l-70.3 77.7a8.3 8.3 0 00-2.1 5.4V592c0 2.2 1.8 4 4 4h116c2.2 0 4-1.8 4-4v-36c0-2.2-1.8-4-4-4h-68.4l70.3-77.7a8.3 8.3 0 002.1-5.4V432c0-2.2-1.8-4-4-4z"
          />
        </svg>
      </span>
    </button>
  </h4>
  <div
    style={
      Object {
        "clear": "both",
      }
    }
  />
</div>
`;

exports[`invalid icon to button 1`] = `
<div>
  <h4
    className="table-title-heading"
  >
    Title Of Table
  </h4>
  <h4
    className="table-title-button"
  >
    <button
      className="ant-btn"
      label="Button With Callbac"
      onClick={[Function]}
      type="button"
    >
      <asd />
      <span>
        Button With Callbac
      </span>
    </button>
  </h4>
  <div
    style={
      Object {
        "clear": "both",
      }
    }
  />
</div>
`;

exports[`render fails if no title prop given 1`] = `
<div>
  <h4
    className="table-title-heading"
  />
  
  <div
    style={
      Object {
        "clear": "both",
      }
    }
  />
</div>
`;

exports[`renders title prop and basic button 1`] = `
<div>
  <h4
    className="table-title-heading"
  >
    Title Of Table
  </h4>
  <h4
    className="table-title-button"
  >
    <button
      className="ant-btn"
      label="My Button"
      onClick={[Function]}
      type="button"
    >
      <span>
        My Button
      </span>
    </button>
  </h4>
  <div
    style={
      Object {
        "clear": "both",
      }
    }
  />
</div>
`;

exports[`renders title prop and button with onClick handler 1`] = `
<div>
  <h4
    className="table-title-heading"
  >
    Title Of Table
  </h4>
  <h4
    className="table-title-button"
  >
    <button
      className="ant-btn"
      label="Button With Callbac"
      onClick={[Function]}
      type="button"
    >
      <span>
        Button With Callbac
      </span>
    </button>
  </h4>
  <div
    style={
      Object {
        "clear": "both",
      }
    }
  />
</div>
`;

exports[`renders title prop and not the button 1`] = `
<div>
  <h4
    className="table-title-heading"
  >
    Title Of Table
  </h4>
  
  <div
    style={
      Object {
        "clear": "both",
      }
    }
  />
</div>
`;

exports[`renders without crashing 1`] = `
<div>
  <h4
    className="table-title-heading"
  >
    Title Of Table
  </h4>
  
  <div
    style={
      Object {
        "clear": "both",
      }
    }
  />
</div>
`;
